p2p/transport/websocket/real_ip.go

```diff
@@ -0,0 +1,79 @@
+package websocket
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+func GetRealIP(addr net.Addr, h http.Header) string {
+	remoteAddr := addr.String()
+	remoteIp := GetRealIPFromHeader(h)
+	if remoteIp != nil {
+		remoteTcpAddr, ok := addr.(*net.TCPAddr)
+		if ok {
+			remoteAddr = IpPort(remoteIp, strconv.Itoa(remoteTcpAddr.Port))
+		} else {
+			_, port, err := net.SplitHostPort(remoteAddr)
+			if err == nil {
+				remoteAddr = IpPort(remoteIp, port)
+			}
+		}
+	}
+	return remoteAddr
+}
+
+// GetRealIPFromHeader extracts the client's real IP address from HTTP request header.
+// It checks various proxy header to find the actual IP.
+func GetRealIPFromHeader(h http.Header) net.IP {
+	// Check X-Real-IP header (used by Nginx and others)
+	ipStr := h.Get("X-Real-IP")
```

Comment:
This function seems hacky. It doesn't seem like this is standardized. It might be preferable to force the user to define this function rather than try solving all cases.

One problem here is what happens if my reverse proxy is only built to handle one of these header keys, and the client can smuggle a different value.


For example, imagine if my proxy only knows about "True-Client-IP" header, but it doesn't know it should strip "X-Real-IP", "X-Forwarded-For", or "CF-Connecting-IP". That means that a client can simply set one of those other headers and trick the server.

Reply:
Ah the standard seems to be: https://datatracker.ietf.org/doc/html/rfc7239

I think it makes sense to support this standard. But allow users to override the standard behavior sense it seems not all proxies support this standard(?)

Reply:
The default is rfc7239 only, but I modified it to be customizable.